ncursesw patch

sjpp opened the pull request:
Hi,
I maintain logtop package for openSUSE and got a patch submitted (which I accepted) to use ncursesw instead of old ncurses only. This may imply to drop support for old ncurses releases that have no pkgconfig scripts at all.

Commit:
applied ncursesw patch

JulienPalard commented:
TBH I'm discovering ncursesw thanks to this issue.

Does using ncursesw fixes something in logtop?
Does it breaks something in logtop? I'm currently happy with logtop being able to blindly handle bytes without caring about encoding, is using ncursesw bring encoding problems?

I can understsand that using ncursesw make sense in an application displaying boxes, buttons, and so on, to get proper handling maybe of character width and counts, for alignments and so on. But I'm only using mvwprintw, werase and wrefresh here.

sogalgeeko commented:
Hi,
Actually no, it does not "fix" anything, logtop does not need be fixed :)
It simply improves it by adding a complete UTF-8 support which tend to be, AFAIK, a standard nowadays. In the builds I have made, this patch does not break anything in logtop behavior but I can speak only for openSUSE builds. Maybe it will on some terminals that does no handle non-ASCII characters properly.
As a maintainer it is my duty to bring the patch upstream but, TBH, I can tell you anything else than : it builds and works perfectly in openSUSE releases :)

JulienPalard commented:
OK two surprises for me here.

First I though (but I was wrong, I did not worked on logotp since a few years) that I was only getting bytes in and bytes out so I was compatible with any encoding as I was *not* trying to decode.

I was wrong:

```
$ python3 -c 'import unicodedata; print("\n".join(chr(x) for x in range(130000) if "HEART" in unicodedata.name(chr(x), "")))' | logtop
45 elements
   1    1 ...
   2    1 ...
   3    1 ...
   4    1 ...
   5    1 ...
   6    1 ...
   7    1 ...
   8    1 ...
   9    1 ...
  10    1 ...
```

So I became interested by your proposal, and tried it:

```
$ python3 -c 'import unicodedata; print("\n".join(chr(x) for x in range(130000) if "HEART" in unicodedata.name(chr(x), "")))' | ./logtop
45 lines, 45.00 lines/s
RANK    CNT   LINE/S LINE
   1      1     1.00 ...                                                                                                                           
   2      1     1.00 ...                                                                                                                           
   3      1     1.00 ...                                                                                                                           
   4      1     1.00 ...                                                                                                                           
   5      1     1.00 ...                                                                                                                           
   6      1     1.00 ...                                                                                                                           
   7      1     1.00 ...                                                                                                                           
   8      1     1.00 ...                                                                                                                           
   9      1     1.00 ...                                                                                                                           
  10      1     1.00 ...                     
```

Sadly, it didn't helped, looks like we'll have to take a look further to fix this... I'm interested now!

JulienPalard committed:
Allow UTF-8 to pass thrue.

So typically:

$ make; python3 -c 'import unicodedata; print("\n".join(chr(x) for x in range(130000) if "HEART" in unicodedata.name(chr(x), "")))' | ./logtop
make: 'logtop' is up to date.
45 lines, 45.00 lines/s
RANK    CNT   LINE/S LINE
   1      1     1.00 ☙
   2      1     1.00 ♡
   3      1     1.00 ♥
   4      1     1.00 ❣
   5      1     1.00 ❤
   6      1     1.00 ❥
   7      1     1.00 ❦
   8      1     1.00 ❧
   9      1     1.00 ⺖
  10      1     1.00 ⺗

JulienPalard commented:
Hi @sogalgeeko, I pushed to your PR, so we now get:

```
$ make; python3 -c 'import unicodedata; print("\n".join(chr(x) for x in range(130000) if "HEART" in unicodedata.name(chr(x), "")))' | ./logtop
make: 'logtop' is up to date.
45 lines, 45.00 lines/s
RANK    CNT   LINE/S LINE
   1      1     1.00 ☙                                                                                                                           
   2      1     1.00 ♡                                                                                                                           
   3      1     1.00 ♥                                                                                                                           
   4      1     1.00 ❣                                                                                                                           
   5      1     1.00 ❤                                                                                                                           
   6      1     1.00 ❥                                                                                                                           
   7      1     1.00 ❦                                                                                                                           
   8      1     1.00 ❧                                                                                                                           
   9      1     1.00 ⺖                                                                                                                           
  10      1     1.00 ⺗                    
```

what do you think?

sogalgeeko commented:
Hi!
Looks great! I'll update the package with your commit.
Thank you :)

JulienPalard commented:
Hi @sogalgeeko, I pushed to your PR, so we now get:

```
$ make; python3 -c 'import unicodedata; print("\n".join(chr(x) for x in range(130000) if "HEART" in unicodedata.name(chr(x), "")))' | ./logtop
make: 'logtop' is up to date.
45 lines, 45.00 lines/s
RANK    CNT   LINE/S LINE
   1      1     1.00 ☙                                                                                                                           
   2      1     1.00 ♡                                                                                                                           
   3      1     1.00 ♥                                                                                                                           
   4      1     1.00 ❣                                                                                                                           
   5      1     1.00 ❤                                                                                                                           
   6      1     1.00 ❥                                                                                                                           
   7      1     1.00 ❦                                                                                                                           
   8      1     1.00 ❧                                                                                                                           
   9      1     1.00 ⺖                                                                                                                           
  10      1     1.00 ⺗                    
```

what do you think?
